### PR TITLE
Add logsumexp op

### PR DESCRIPTION
**Pull request by JackCaoG**

### CODE REVIEW

File: torch_xla/csrc/aten_xla_type.cpp [hunk 1]
```diff
@@ -1571,6 +1571,14 @@ std::tuple<at::Tensor, at::Tensor> AtenXlaType::log_sigmoid_forward(
                          bridge::AtenFromXlaTensor(std::get<1>(result_tuple)));
 }
 
+at::Tensor AtenXlaType::logsumexp(const at::Tensor& self, at::IntArrayRef dim,
+                                  bool keepdim) {
+  XLA_FN_COUNTER("xla::");
+  return bridge::AtenFromXlaTensor(XLATensor::logsumexp(
+      bridge::GetXlaTensor(self), xla::util::ToVector<xla::int64>(dim),
+      /*keep_reduced_dimensions*/ keepdim));
```
Comment (dlibenzi) [inline torch_xla/csrc/aten_xla_type.cpp]: `/*keep_reduced_dimensions=*/`

File: torch_xla/csrc/aten_xla_type.h [hunk 2]
```diff
@@ -501,6 +501,9 @@ class AtenXlaType {
   static std::tuple<at::Tensor, at::Tensor> log_sigmoid_forward(
       const at::Tensor& self);
 
+  static at::Tensor logsumexp(const at::Tensor& self, at::IntArrayRef dim,
+                              bool keepdim);
+
   static at::Tensor logdet(const at::Tensor& self);
```
Comment (dlibenzi) [inline torch_xla/csrc/aten_xla_type.h]: Can you re-order `logdet()` given that you are around here?
Should be moved up some place, to keep alphabetically ordered.

File: torch_xla/csrc/tensor_methods.cpp [hunk 3]
```diff
@@ -1399,6 +1400,16 @@ XLATensor XLATensor::log_softmax_backward(const XLATensor& grad_output,
       grad_output.GetIrValue(), output.GetIrValue(), dim));
 }
 
+XLATensor XLATensor::logsumexp(const XLATensor& input,
```
Comment (dlibenzi) [inline torch_xla/csrc/tensor_methods.cpp]: This should be moved down. Functions (which are ops implementation) must be sorted alphabetically.

File: torch_xla/csrc/tensor.h [hunk 4]
```diff
@@ -602,6 +602,10 @@ class XLATensor {
                                         const XLATensor& output,
                                         xla::int64 dim);
 
+  static XLATensor logsumexp(const XLATensor& input,
```
Comment (dlibenzi) [inline torch_xla/csrc/tensor.h]: Proper order.

File: torch_xla/csrc/reduction.cpp [hunk 5]
```diff
@@ -422,4 +429,25 @@ xla::XlaOp BuildAny(xla::XlaOp input, absl::Span<const xla::int64> dimensions,
...
+                        .result;
+  xla::XlaOp logs = xla::Log(sums);
+  // If keep_reduced_dimensions is false, we need to reshpae the maxInDim to the
+  // reduced shape before doing the add.
+  if (!keep_reduced_dimensions) {
+    maxInDim = CreateSummation(maxInDim, dimensions, keep_reduced_dimensions,
+                               /*scale=*/false)
+                   .result;
+  }
+  return logs + maxInDim;
```
Comment (dlibenzi) [inline torch_xla/csrc/reduction.cpp]: `max_in_dim`

File: torch_xla/csrc/ops/logsumexp.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+#include "torch_xla/csrc/reduction.h"
+#include "torch_xla/csrc/tensor_util.h"
+#include "torch_xla/csrc/torch_util.h"
+
+namespace torch_xla {
+namespace ir {
+namespace ops {
+namespace {
+
+xla::XlaOp LowerLogsumexp(xla::XlaOp input,
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/logsumexp.cpp]: No need for this wrapper.

File: torch_xla/csrc/tensor_methods.cpp (hunk 3 above)
Comment (JackCaoG) [reply torch_xla/csrc/tensor_methods.cpp]: Yup, I first name this log_sum_exp and forgot to reorder functions after rename it.

**Pull request closed without merging**